crates/hir-ty/src/method_resolution.rs

```diff
@@ -216,7 +259,30 @@ impl TraitImpls {
             for konst in module_data.scope.unnamed_consts() {
                 let body = db.body(konst.into());
                 for (_, block_def_map) in body.blocks(db.upcast()) {
-                    Self::collect_def_map(db, map, &block_def_map);
+                    Self::collect_def_map(db, map, &block_def_map, query_depth);
+                }
+            }
+
+            if matches!(query_depth, ImplQueryDepth::SurfaceOnly) {
+                continue;
+            }
+
+            // Collect nested impls inside function bodies, etc.
+            for (_, ns) in module_data.scope.entries() {
+                for module in ns.types.iter().map(|t| t.0).chain(ns.values.iter().map(|t| t.0)) {
+                    let body: Option<DefWithBodyId> = match module {
+                        ModuleDefId::FunctionId(it) => Some(it.into()),
+                        ModuleDefId::EnumVariantId(it) => Some(it.into()),
+                        ModuleDefId::ConstId(it) => Some(it.into()),
+                        ModuleDefId::StaticId(it) => Some(it.into()),
+                        _ => None,
+                    };
+                    if let Some(body) = body {
+                        let body = db.body(body);
+                        for (_, block_def_map) in body.blocks(db.upcast()) {
+                            Self::collect_def_map(db, map, &block_def_map, query_depth);
+                        }
+                    }
                 }
             }
```

Comment:
The logic is quite verbose, but I couldn't find any good ways that can probe into those blocks 🤔